src/rpc/api/validator/validator.ts

```diff
@@ -37,46 +49,54 @@ export class ValidatorApi implements IValidatorApi {
     this.opPool = opPool;
   }
 
-  public async getDuties(validatorIndex: ValidatorIndex): Promise<{currentVersion: Fork; validatorDuty: ValidatorDuty}> {
-    // eslint-disable-next-line @typescript-eslint/no-object-literal-type-assertion
-    return {} as {currentVersion: Fork; validatorDuty: ValidatorDuty};
-  }
-
-  public async produceBlock(slot: Slot, randaoReveal: bytes): Promise<BeaconBlock> {
-    const block: BeaconBlock = {
-      body: undefined,
-      previousBlockRoot: undefined,
-      signature: undefined,
-      slot: undefined,
-      stateRoot: undefined
-    };
-    const prevBlock: BeaconBlock = await this.db.getChainHead();
-    const curState: BeaconState = await this.db.getState();
-    // Note: To calculate state_root,
-    // the validator should first run the state transition function on an unsigned block
-    // containing a stub for the state_root.
-    // It is useful to be able to run a state transition function that does not
-    // validate signatures or state root for this purpose.
-    block.slot = slot;
-    block.previousBlockRoot = ssz.hashTreeRoot(prevBlock, BeaconBlock);
-    block.stateRoot = ssz.hashTreeRoot(curState, BeaconState);
-    // TODO Eth1Data
-    return block;
+  public async produceBlock(slot: Slot, randaoReveal: bytes96): Promise<BeaconBlock> {
+    return await assembleBlock(this.db, this.opPool, slot, randaoReveal);
   }
 
   public async isProposer(index: ValidatorIndex, slot: Slot): Promise<boolean> {
     const state: BeaconState = await this.db.getState();
     return isProposerAtSlot(state, slot, index);
   }
 
+  public async getDuties(validatorPublicKeys: Buffer[]): Promise<ValidatorDuty[]> {
+    const state = await this.db.getState();
+
+    const validatorIndexes = await Promise.all(validatorPublicKeys.map(async publicKey => {
+      return  await this.db.getValidatorIndex(publicKey);
+    }));
+
+    const proposerIndex = getBeaconProposerIndex(state);
+    const validatorProposer = validatorIndexes.indexOf(proposerIndex);
+
+    return validatorPublicKeys.map(
+      this.assembleValidatorDuty(
```

Comment:
It might be nice to pull out some of the duty and attestation assembling in the same way that block assembling was pulled out into separate files.
Maybe they can all live in a new module together. eg: `src/validatorDuties` or something like that.

Reply:
Yeah, agree, I was thinking something like `chain/factory/block` and `chain/factory/attestation`.